torch/utils/data/datapipes/utils/decoder.py

```diff
@@ -228,29 +222,47 @@ def torch_audio(key, data):
             return torchaudio.load(fname)
 
 
-
 ################################################################
-# a sample decoder
+# mat
 ################################################################
+class MatHandler:
+    def __init__(self, **loadmat_kwargs) -> None:
+        try:
+            import scipy.io as sio
+        except ImportError as e:
+            raise ModuleNotFoundError("Package `scipy` is required to be installed for mat file."
+                                      "Please use `pip install scipy` or `conda install scipy`"
+                                      "to install the package")
+        self.sio = sio
+        self.loadmat_kwargs = loadmat_kwargs
+
+    def __call__(self, extension, data):
+        if extension != 'mat':
+            return None
+        with io.BytesIO(data) as stream:
+            return self.sio.loadmat(stream, **self.loadmat_kwargs)
 
 
+################################################################
+# a sample decoder
+################################################################
 class Decoder:
     """
     Decode key/data sets using a list of handlers.
     For each key/data item, this iterates through the list of
     handlers until some handler returns something other than None.
     """
 
-    def __init__(self, handlers):
-        self.handlers = handlers
+    def __init__(self, handlers, key_fn):
+        self.handlers = list(handlers) if len(handlers) > 0 else []
+        self.key_fn = key_fn
 
+    # Add from the beginning of the handlers to make sure the added
+    # handler having highest priority
     def add_handler(self, handler):
         if not handler:
             return
-        if not self.handlers:
-            self.handlers = [handler]
-        else:
-            self.handlers.append(handler)
+        self.handlers = [handler] + self.handlers
```

Comment:
This avoids creating a new list over and over.

```suggestion
        self.handlers = self.handlers.insert(0, handler)
```